site/_posts/2017-09-27-ray-0.2-release.markdown

```diff
@@ -0,0 +1,148 @@
+15GB/s for large objects (when writing from a single client). Achieving this
+performance requires enabling huge pages (to minimize the number of TLB cache
+misses). Instructions for doing so are [here][1].
+
+The speed at which objects can be written into the object store is a key
+performance metric. For example, this is the bottleneck for [A3C][6] and many
+other algorithms.
+
+You can benchmark write throughput as follows.
+
+```python
+import numpy as np
+import ray
+
+ray.init()
+
+x = np.ones(10 ** 9, dtype=np.uint8)
+
+# Measure the time required to write 1GB to the Plasma store.
+%time x_id = ray.put(x)
+```
+
+## Web UI
+
+We've built an initial Jupyter-notebook-based web UI for understanding and
+debugging application performance. See the [instructions for using the UI][2].
+The UI includes a task timeline visualization based on Chrome tracing to see
+where tasks were scheduled, how long they took, and what the dependencies
+between the tasks were. An example visualization is shown below.
+
+<div align="center">
+<img src="{{ site.base-url }}/assets/ray_0.2_release/timeline_visualization.png">
+</div>
+<div><i>A visualization of the task timeline. Boxes indicate tasks, and the
+different colors indicate the time spent executing the task versus storing the
+the results in the object store. Here arrows indicate data dependencies between
+tasks.</i></div>
+<br />
+
+This type of visualization can immediately expose problems with performance,
+scheduling, and load balancing.
+
+The above visualization can be generated on a single machine by the following
+script.
+
+```python
+import ray
+import time
+
+ray.init()
+
+@ray.remote
+def f(x):
+    time.sleep(0.001)
+    return 1
+
+@ray.remote
+def g(*ys):
+    return 1
+
+time.sleep(1)
+
+x = 1
+for _ in range(3):
+    ys = [f.remote(x) for _ in range(8)]
+    x = g.remote(*ys)
+```
+
+## RLlib
+
+We've begun implementing a [scalable reinforcement learning library][3] based on
+Ray. So far it includes implementations of the following algorithms.
+
+- Proximal policy optimization (PPO)
+- Deep Q-learning (DQN)
+- Asynchronous advantage actor critic (A3C)
+- Evolution Strategies (ES)
+
+The DQN, A3C, and ES implementations are based on the [OpenAI baselines][7].
+Example code for training is available and can be used as follows.
+
+```
+# On a single machine.
+python ray/python/ray/rllib/train.py --alg=PPO --\
+                                     --env=Humanoid-v1
+
+# On a cluster.
+python ray/python/ray/rllib/train.py --alg=PPO \
+                                     --env=Humanoid-v1 \
+                                     --redis-address=<head-node-ip>:6379
+```
+
+This uses [proximal policy optimization][11] to train a policy to control a walking
+humanoid robot in the Mujoco physics engine.
+
+Running this on AWS with a cluster of fifteen m4.16xlarge instances and one
+p2.16xlarge instance, we achieve a reward of over 6000 in around 35 minutes. The
```

Comment:
update this with new results?

Reply:
I think these are the new results, aren't they?

Reply:
oh right, my bad